### ISSUE TITLE
Alias and display name are not editable for custom identity providers

### ISSUE THREAD

**Issue by florianschiller** [keycloak/keycloak#24710]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

identity-brokering

### Describe the bug

My team and I are developing a custom identity provider which can be found [here](https://github.com/Bundesdruckerei-GmbH/Gematik-IDP). Since it was nearly impossible to provide a UI for a custom identity provider as part of an open source release in the past, we were happy to try out the new dynamic identity provider UI feature introduced with Keycloak 22. Unfortunately, it is not possible to set the display name and alias fields via this new feature, as these are direct variables of ``IdentityProviderModel`` and not part of the ``config`` map. The behaviour of  ``DynamicComponents.convertToName`` makes it impossible to add a field to the frontend for a variable which isn't part of that map.

Display name and alias are shown for custom identity providers which contain ``-oidc`` in their provider-id, but this providers don't support adding a custom UI. In our case, using such a provider ID does not make sense because our other configuration fields could not be added.

### Version

22.0.5

### Expected behavior

Either:

- It is possible to add a display name and alias field to a dynamic identity provider UI
- Or display name and alias are default fields of each identity provider like display order as well as client id and client secret

### Actual behavior

Display name and alias cannot be added as seperate fields to a custom identity provider

### How to Reproduce?

You can reproduce the bug when configuring a custom identity provider with the following dynamic UI:
````java
    @Override
    public List<ProviderConfigProperty> getConfigProperties() {
        return ProviderConfigurationBuilder
            .create().property()
            .name("alias")
            .label("Alias")
            .helpText(
                "The alias uniquely identifies an identity provider and it is also used to build the redirect uri.")
            .type(ProviderConfigProperty.STRING_TYPE)
            .add()
            .property()
            .name("displayName")
            .label("Display Name")
            .helpText("Friendly name for Identity Providers.")
            .type(ProviderConfigProperty.STRING_TYPE)
            .add()
            .build();
    }
````
In this case, alias is saved a ``config.alias`` and display name as ``config.displayName``.

### Anything else?

There is already a discussion on a similar topic [here](https://github.com/keycloak/keycloak/pull/24266#issuecomment-1793439822), but it was about removing the standard fields Client ID and Client Secret.

**Comment by jonkoops** [comment]
I think this is certainly a valid concern I think, and a limitation of the current API. Do you have a proposal on how you would like to see this solved?

@ssilvert @edewit I think this might be up your alley.

**Comment by ssilvert** [comment]
> I think this is certainly a valid concern I think, and a limitation of the current API. Do you have a proposal on how you would like to see this solved?
> 
> @ssilvert @edewit I think this might be up your alley.

I agree.  Assigning to @edewit so we can get his take on it before we move it out of triage.

**Comment by florianschiller** [comment]
> I think this is certainly a valid concern I think, and a limitation of the current API. Do you have a proposal on how you would like to see this solved?
> 
> @ssilvert @edewit I think this might be up your alley.

As changing the behaviour of ``DynamicComponents.convertToName`` would affect a multitude of custom plugins and would make them not  backwards compatible, I would suggest adding display name and alias as default fields to ``AddIdentityProvider`` and ``DetailSettings``.

**Comment by edewit** [comment]
It's meant to be all stored in `config` changing `convertToName` wouldn't fix that. Adding display name and alias to those pages also wouldn't solve it as all the social Idp don't have these fields

**Comment by edewit** [comment]
You can work around it by overriding the `create` and remove these fields from the config and add them to the root

```java
T create(KeycloakSession session, ComponentModel model);
```
like this:
```java
    @Override
    public GematikIDP create(KeycloakSession session, ComponentModel model) {
        final String displayName = model.getConfig().remove("displayName").get(0);
```

**Comment by ssilvert** [comment]
@florianschiller Are you OK with closing this?

**Comment by florianschiller** [comment]
Hello @ssilvert ,

I looked into it and it seems like adding the mentioned workaround to the `create` function doesn't fix the issue. I see two problems with it:

1. As far as I was able to determine, the alias and display name where not saved into the identity provider model returned by the Keycloak-API after I integrated this changes both in the `create` function as well as the `GematikIDPConfig` via an init block. I was unable to see any difference in the frontend after changing the display name.
2. Even if it would be working, I would instead use a `get` instead of a `remove` here. If both `alias` and `displayName` are being removed from the `config` and are therefore left empty in the frontend in future calls, saving the identity provier would empty the root fields on consecutive save requests if I'm not mistaken. This is especially a problem with the `alias` field, since it is used as part of the ID in the database. 

In my opinion, this seems a little bit to hacky to be implemented in a production ready identity provider plugin. If this is the only solution to this problem, we most likely will not integrate it and just leave it as it is.

**Comment by edewit** [comment]
@florianschiller yes your right it is a bit hacky, but why do you absolutely need these fields at a top level? Couldn't it just be config elements?

**Comment by florianschiller** [comment]
Hello @edewit ,

For us, it doesn't make a difference if this two fields are part of the top-level or the config map. Since the alias is used as the IDP ID, we have to define it at the time of creation and it should be immutable afterwards. Additionally, it is required to be set at the top-level by Keycloak itself and not by our requirements.

The later part holds thruth for the display name, too. The default Keycloak UI needs it to be defined on the top-level, but we are currently unable to do so with the currently implementation of the dynamic configuration page.

Like I said before, the core problem is that a developer is only able to define configuration parameters as part of the `config` map, when they implement a custom dynamic configuration page. This means, alias and display name aren't definable for custom dynamic configuration page, which leads to the problems mentioned above.

**Comment by edewit** [comment]
after a dumb attempt to fix this I dived more into this and I now know how it works the alias seems to always be the providerId, in your case 'gematik-idp' so you don't need to put it in the config.

**Comment by florianschiller** [comment]
Yes, but if I want to have multiple IDP's of the same kind, I need to be able to change the alias since it is used as the ID of the IDP.

**Comment by nosnilmot** [comment]
> Display name and alias are shown for custom identity providers which contain `-oidc` in their provider-id, but this providers don't support adding a custom UI.

This is no longer true. Since https://github.com/keycloak/keycloak/pull/25697 (Keycloak 24+) custom identity providers with `oidc` in their provider-id can add custom fields to the UI too!

**Comment by rtufisi** [comment]
Hei [nosnilmot](https://github.com/nosnilmot),

I've tried your approach. I've created  `CustomOIDCProviderFactory`  which extends `AbstractIdentityProviderFactor` . I've given it a `PROVIDER_ID = "custom-provider-oidc"`.
After I create a new identity provider if this custom type I can see the ` OpenID Connect settings` section and  some fields specific to the `OpenID Connect provider`. The `displayName` is editable and works but the `alias` is still hardcoded. 
I've tried adding it as a `DynamicComponent` but it didn't work.

I there a workaround to have a editable `alias`?
